open_seq2seq/utils/utils.py

```diff
@@ -439,4 +449,301 @@ def cast_types(input_dict, dtype):
+                        "You should probably not provide "
+                        "\"--continue_learning\" flag?")
+        checkpoint = None
+    elif (args.mode == 'infer' or args.mode == 'eval' or
+        args.mode == 'interactive_infer'):
+      if os.path.isdir(logdir) and os.listdir(logdir) != []:
+        checkpoint = tf.train.latest_checkpoint(ckpt_dir)
+        if checkpoint is None:
+          raise IOError(
+              "There is no valid TensorFlow checkpoint in the "
+              "{} directory. Can't load model".format(ckpt_dir)
+          )
+      else:
+        raise IOError(
+            "{} does not exist or is empty, can't restore model".format(
+                ckpt_dir
+            )
+        )
+  except IOError as e:
+    if args.no_dir_check:
+      print("Warning: {}".format(e))
+      print("Resuming operation since no_dir_check argument was provided")
+    else:
+      raise
+
+  return checkpoint
+
+def create_logdir(args, base_config):
+  """A helper function that ensures the logdir and log files are setup corretly.
+  Only called in --enable_logs is set.
+
+   Args:
+    args (dict): Dictionary as returned from get_base_config()
+    base_config (dict): Dictionary as returned from get_base_config()
+
+  Returns:
+    Some objects that need to be cleaned up in run.py
+  """
+  logdir = base_config['logdir']
+  if not os.path.exists(logdir):
+    os.makedirs(logdir)
+
+  tm_suf = datetime.datetime.now().strftime('%Y-%m-%d_%H-%M-%S')
+  shutil.copy(
+      args.config_file,
+      os.path.join(logdir, 'config_{}.py'.format(tm_suf)),
+  )
+
+  with open(os.path.join(logdir, 'cmd-args_{}.log'.format(tm_suf)),
+            'w') as f:
+    f.write(" ".join(sys.argv))
+
+  with open(os.path.join(logdir, 'git-info_{}.log'.format(tm_suf)),
+            'w') as f:
+    f.write('commit hash: {}'.format(get_git_hash()))
+    f.write(get_git_diff())
+
+  old_stdout = sys.stdout
+  old_stderr = sys.stderr
+  stdout_log = open(
+      os.path.join(logdir, 'stdout_{}.log'.format(tm_suf)), 'a', 1
+  )
+  stderr_log = open(
+      os.path.join(logdir, 'stderr_{}.log'.format(tm_suf)), 'a', 1
+  )
+  sys.stdout = Logger(sys.stdout, stdout_log)
+  sys.stderr = Logger(sys.stderr, stderr_log)
+
+  return old_stdout, old_stderr, stdout_log, stderr_log
+
+def create_models(args, base_config, config_module, base_model, hvd):
+  """A helpful function that creates the train, eval, and infer models as
+  needed.
+
+  Args:
+    args (dict): Dictionary as returned from get_base_config()
+    base_config (dict): Dictionary as returned from get_base_config()
+    config_module: config_module as returned from get_base_config()
+    base_model (OpenSeq2Seq model): Dictionary as returned from
+      get_base_config()
+    hvd: Either None if Horovod is not enabled, or the Horovod library
+
+  Returns:
+    models (list): A 3 element list consisting of the train, eval, and infer
```

Comment:
In my opinion it is better to just return whatever model was created. This will be either one model or tuple of two in case of train_eval mode. This way a user would not need to remember in which order the modes go into this returned list.